spec.md

```diff
@@ -0,0 +1,1597 @@
+* `as J[] from (relation) A relates I[]` removes $`I <_! J`$
+
+**Case ATT**
+* `attribute A` removes $`A : \mathbf{Att}`$ and $`A : \mathbf{Att}(O_A)`$
+* `value V from (attribute) A value V` removes $`A < V`$
+* `sub B from (attribute) A` removes $`A <_! B`$ and $`O_A <_! O_B`$
+
+**Case PLAYS**
+* `plays B:I from (type) A` removes $`A <_! I`$ 
+
+**Case OWNS**
+* `owns B from (type) A` removes $`A <_! O_B`$ 
+* `owns B[] from (type) A` removes $`A <_! O_B`$
+
+### Constraints
+
+_In each case, `undefine` removes the postulated condition (restoring the default)._ (minor exception: subkey)
+
+**Case CARD**
+* `@card(n..m) from A relates I`
+* `@card(n..m) from A plays B:I`
+* `@card(n...m) from A owns B`
+
+**Case CARD_LIST**
+* `@card(n..m) from A relates I[]`
+* `@card(n...m) from A owns B[]`
+
+**Case PLAYS_AS**
+* `as C from A plays B`
+
+**Case OWNS_AS**
+* `as C from A owns B`
+
+**Case UNIQUE**
+* `@unique from A owns B`
+
+**Case KEY**
+* `@key from A owns B`
+
+**Case SUBKEY**
+* `@subkey(<LABEL>) from A owns B` removes $`B`$ as part of the `<LABEL>` key of $`A`$
+
+**Case ABSTRACT**
+* `@abstract from (type) B` 
+* `@abstract from A plays B:I`
+* `@abstract from A owns B` 
+* `@abstract from A owns B[]` 
+* `@abstract from B relates I`
+* `@abstract from B relates I[]`
+
+**Case VALUES**
+* `@values(v1, v2) from A owns B` 
+* `@range(v1..v2) from A owns B`
+* `@values(v1, v2) from A values B` 
+* `@range(v1..v2) from A values B`
+
+**Case DISTINCT**
+* `@distinct from A owns B[]`
+* `@distinct from B relates I[]`
+
+### Triggers
+
+_In each case, `undefine` removes the triggered action._
+
+**Case DEP_DEL (CASCADE/INDEPEDENT)**
+* `@cascade from (relation) B relates I`
+* `@cascade from (relation) B`
+* `@independent from (attribute) B`
+
+### Value types
+
+**Case PRIMITIVES**
+cannot undefine primitives
+
+**Case STRUCT**
+
+* `struct S;`
+  removes $S : \mathbf{Type}$ and all associated defs.
+  * **TT** error if
+    * $S$ is used in another struct
+    * $S$ is used as value type of an attribute
+
+### Functions defs
+
+**Case STREAM_RET_FUN**
+* `fun F;`
+  removes $F$ and all associated defs.
+  * **TT** error if
+    * $S$ is used in another function
+
+
+**Case SINGLE_RET_FUN**
+* `fun f;`
+  removes $f$ and all associated defs.
+  * **TT** error if
+    * $S$ is used in another function
+
+_Comment: notice difference in capitalization between the two cases!_
+
+## Redefine semantics
+
+_In each case, redefine acts like an undefine (which cannot be a no-op) and a define of the given axiom_
+
+**Redefine principles**: 
+
+1. Can only redefine one thing at a time
+2. We disallow redefining boolean properties:
+  * _Example 1_: a type can either exists or not. we cannot "redefine" it's existence, but only define or undefine it.
+  * _Example 2_: a type is either abstract or not. we can only define or undefine `@abstract`.
+
+### Type axioms
+
+**Case ENT**
+* cannot redefine `entity A`
+* `(entity) A sub B` redefines $`A < B`$
```

Comment:
This could be more specific, especially as redefinitions are not commonplace in DDLs. Perhaps something like:
```suggestion
* `(entity) A sub B` undefines $`A <_! C`$ and defines $`A <_! B`$ where previously $`A <_! C`$.
```
Especially as redefines cannot be no-ops (if I've understood correctly), and the axioms being undefined cannot be determined from the redefine syntax, it is necessary to specify those axioms to make apparent the conditions under which a no-op occurs. This applies to all redefines.